File: src/backend/app/models/AssessmentToken.java
package models;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
@Table(name = "assessment_token")
public class AssessmentToken {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "risk_assessment_id", nullable = false)
    @NotNull
    private RiskAssessment riskAssessment;

    @Column(name = "token", nullable = false, unique = true, length = 255)
    @NotNull
    private String token;

    @Column(name = "respondent_email", nullable = false, length = 255)
    @NotNull
    private String respondentEmail;

    @Column(name = "expires_at", nullable = false)
    @NotNull
    private LocalDateTime expiresAt;

    @Column(name = "used_at")
    private LocalDateTime usedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    // Constructors
    public AssessmentToken() {}

    public AssessmentToken(RiskAssessment riskAssessment, String token, String respondentEmail, LocalDateTime expiresAt) {
        this.riskAssessment = riskAssessment;
        this.token = token;
        this.respondentEmail = respondentEmail;
        this.expiresAt = expiresAt;
    }

    // Helper methods
    public boolean isExpired() {
        return LocalDateTime.now().isAfter(expiresAt);
    }

    public boolean isUsed() {
        return usedAt != null;
    }

    public boolean isValid() {
        return !isExpired() && !isUsed();
    }

    public void markAsUsed() {
        this.usedAt = LocalDateTime.now();
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public RiskAssessment getRiskAssessment() {
        return riskAssessment;
    }

    public void setRiskAssessment(RiskAssessment riskAssessment) {
        this.riskAssessment = riskAssessment;
    }

    public String getToken() {
        return token;
    }

    public void setToken(String token) {
        this.token = token;
    }

    public String getRespondentEmail() {
        return respondentEmail;
    }

    public void setRespondentEmail(String respondentEmail) {
        this.respondentEmail = respondentEmail;
    }

    public LocalDateTime getExpiresAt() {
        return expiresAt;
    }

    public void setExpiresAt(LocalDateTime expiresAt) {
        this.expiresAt = expiresAt;
    }

    public LocalDateTime getUsedAt() {
        return usedAt;
    }

    public void setUsedAt(LocalDateTime usedAt) {
        this.usedAt = usedAt;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    // Lifecycle methods
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    @Override
    public String toString() {
        return "AssessmentToken{" +
                "id=" + id +
                ", riskAssessmentId=" + (riskAssessment != null ? riskAssessment.getId() : null) +
                ", respondentEmail='" + respondentEmail + '\'' +
                ", expiresAt=" + expiresAt +
                ", usedAt=" + usedAt +
                ", createdAt=" + createdAt +
                '}';
    }
}

File: src/backendng/src/main/kotlin/com/secman/domain/AssessmentToken.kt
package com.secman.domain

import io.micronaut.serde.annotation.Serdeable
import jakarta.persistence.*
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import java.time.LocalDateTime
import java.util.*

@Entity
@Table(name = "assessment_token")
@Serdeable
data class AssessmentToken(
    @Id
    @GeneratedValue
    var id: Long? = null,

    @Column(name = "token", nullable = false, unique = true)
    @NotBlank
    var token: String,

    @Column(name = "email", nullable = false)
    @Email
    @NotBlank
    var email: String,

    @Column(name = "expires_at", nullable = false)
    @NotNull
    var expiresAt: LocalDateTime,

    @Column(name = "is_used")
    var isUsed: Boolean = false,

    @Column(name = "used_at")
    var usedAt: LocalDateTime? = null,

    @ManyToOne
    @JoinColumn(name = "risk_assessment_id", nullable = false)
    @NotNull
    var riskAssessment: RiskAssessment,

    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: LocalDateTime? = null,

    @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime? = null
) {

    companion object {
        /**
         * Generate a secure random token string
         */
        fun generateToken(): String {
            return UUID.randomUUID().toString().replace("-", "")
        }

        /**
         * Create a new assessment token with 30-day expiration
         */
        fun create(email: String, riskAssessment: RiskAssessment): AssessmentToken {
            return AssessmentToken(
                token = generateToken(),
                email = email,
                expiresAt = LocalDateTime.now().plusDays(30),
                riskAssessment = riskAssessment
            )
        }
    }

    @PrePersist
    fun onCreate() {
        val now = LocalDateTime.now()
        createdAt = now
        updatedAt = now
    }

    @PreUpdate
    fun onUpdate() {
        updatedAt = LocalDateTime.now()
    }

    /**
     * Check if token is valid (not expired and not used)
     */
    fun isValid(): Boolean {
        return !isUsed && LocalDateTime.now().isBefore(expiresAt)
    }

    /**
     * Mark token as used
     */
    fun markAsUsed() {
        isUsed = true
        usedAt = LocalDateTime.now()
    }

    override fun toString(): String {
        return "AssessmentToken(id=$id, email='$email', isUsed=$isUsed, expiresAt=$expiresAt)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AssessmentToken) return false
        return id != null && id == other.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}